model/PermissionsService.php

```diff
@@ -54,55 +56,150 @@ public function __construct(
+        core_kernel_classes_Resource $root,
+        array $normalizedRequest,
+        bool $isRecursive
+    ): void {
+        $this->triggerEvents($normalizedRequest, $root->getUri(), $isRecursive);
     }
 
-    public function savePermissions(
-        bool $isRecursive,
-        core_kernel_classes_Class $class,
-        array $privilegesToSet
+    /**
+     * @return Generator<core_kernel_classes_Resource>
+     */
+    public function getNestedResources(core_kernel_classes_Resource $resource): Generator
+    {
+        $stack = new SplStack();
+        $stack->push($resource);
+
+        while (!$stack->isEmpty()) {
+            /** @var $current core_kernel_classes_Resource */
+            $current = $stack->pop();
+            yield $current;
+
+            if ($current->isClass()) {
+                /** @var $current core_kernel_classes_Class */
+                foreach ($current->getSubClasses() as $class) {
+                    $stack->push($class);
+                }
+
+                foreach ($current->getInstances() as $instance) {
+                    yield $instance;
+                }
+            }
+        }
+    }
+
+    private function setPermissionsByResourceRootAndTriggerEvents(
+        core_kernel_classes_Resource $resource,
+        array $normalizedRequest,
+        bool $isRecursive
     ): void {
-        $this->saveResourcePermissions($class, $privilegesToSet, $isRecursive);
+        if (!empty($normalizedRequest)) {
+            $this->applyRequestForResources(
+                $this->getResourcesToUpdate($resource, $isRecursive),
+                $normalizedRequest
+            );
+
+            $this->triggerEvents($normalizedRequest, $resource->getUri(), $isRecursive);
+        }
+    }
+
+    private function applyRequestForResources(array $resources, array $normalizedRequest): void
+    {
+        $permissionsList = $this->getResourcesPermissions($resources);
+        $actions = $this->getActions($resources, $permissionsList, $normalizedRequest);
+
+        $this->dryRun($actions, $permissionsList);
+
+        $this->debug(
+            'applyRequestForResources(): Applying %d operations on %d resources',
+            count($actions['add'] ?? []) + count($actions['remove'] ?? []),
+            count($resources)
+        );
+
+        $this->wetRun($actions);
     }
 
     private function getActions(array $resourcesToUpdate, array $permissionsList, array $addRemove): array
     {
+        $this->debug('getActions(): Handling %d resources', count($resourcesToUpdate));
+
         $actions = ['remove' => [], 'add' => []];
 
+        $i = 0;
         foreach ($resourcesToUpdate as $resource) {
+            //$this->debug('getActions(): Handling resource %d/%d', ++$i, count($resourcesToUpdate));
+
             $currentPrivileges = $permissionsList[$resource->getUri()];
 
             $remove = $this->strategy->getPermissionsToRemove($currentPrivileges, $addRemove);
+            if (count($remove) > 0) {
```

Comment:
question, is there a case where addRemove contains both permissions to add AND remove? :thinking: